pipeline/metadata/flatten_satellite.py

```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
+  """
+
+  # v2.1 measurements repeat 4 times or until the test is successful
+  if not rcodes:
+    successful_test_rcode = []
+    failed_test_rcodes = []
+  elif rcodes[-1] == "0":
+    successful_test_rcode = [rcodes[-1]]
+    failed_test_rcodes = rcodes[0:-1]
+  else:
+    successful_test_rcode = []
+    failed_test_rcodes = rcodes
+
+  return successful_test_rcode, failed_test_rcodes
+
+
+def _process_satellite_v2p1(row: Row, scan: Any,
+                            filepath: str) -> Iterator[Row]:
+  """Finish processing a line of Satellite v2.1 data.
 
   Args:
     row: a partially processed row of satellite data
     scan: a loaded json object containing the parsed content of the line
+    filepath: path like "<path>/<filename>.json.gz"
 
   Yields:
     Rows
   """
-  responses = scan.get('response', [])
-  row['controls_failed'] = not scan['passed_liveness']
-  row['rcode'] = [str(response['rcode']) for response in responses]
-
-  row['average_confidence'] = scan.get('confidence')['average']
-  row['matches_confidence'] = scan.get('confidence')['matches']
-  row['untagged_controls'] = scan.get('confidence')['untagged_controls']
-  row['untagged_response'] = scan.get('confidence')['untagged_response']
-
-  row['excluded'] = scan.get('excluded', False)
-  row['exclude_reason'] = ' '.join(scan.get('exclude_reason', []))
-
-  errors = [
-      response['error']
-      for response in responses
-      if response['error'] and response['error'] != 'null'
-  ]
-  row['error'] = ' | '.join(errors) if errors else None
-  yielded = False
-  has_test_domain = False
-  for response in responses:
-    if response['url'] == row['domain']:
-      has_test_domain = True
-    # Check responses for test domain with valid answers
-    if response['url'] == row['domain'] and (response['rcode'] == 0 and
-                                             response['has_type_a']):
-      row['has_type_a'] = True
-      # Separate into one answer IP per row for tagging
-      row['received'] = []
-      for ip in response['response']:
-        received = {
-            'ip': ip,
-            'http': response['response'][ip].get('http'),
-            'cert': response['response'][ip].get('cert'),
-            'asname': response['response'][ip].get('asname'),
-            'asnum': response['response'][ip].get('asnum'),
-            'matches_control': ''
-        }
-        matched = response['response'][ip].get('matched', [])
-        if matched:
-          received['matches_control'] = ' '.join(
-              [tag for tag in matched if tag in SATELLITE_TAGS])
-        row['received'].append(received)
-      yielded = True
-      yield row.copy()
-    # Do not output rows if there are no trials for the test domain.
-  if not yielded and has_test_domain:
-    # All trials for the test domain in `scan` are errors
-    yield row.copy()
+  row['controls_failed'] = not scan['passed_control']
+  received_ips = scan.get('response').copy()
```

Comment:
Rename `received_ips` to `response`.

Reply:
Actually, maybe we should rename `row` to `base_observation` and `received_ips` to something like `input_response_data`, since it's not a response, and it's not just IPs.

I'm using "observation", instead of "responses" because sometimes the observation is not a response (e.g. timeout)

Reply:
renamed `row` to `base_observation` and `received_ips` to `input_response_data`